### ISSUE TITLE
mvn clean site-deploy with relative path in asciidoc includes in a maven module is not working

### ISSUE THREAD

**Issue by entfrickler** [asciidoctor/asciidoctor-maven-plugin#318]
Hi,
I have a Maven multi module project in which I have integrated the documentation of the modules and a extra module for the user manual. For re usability and for a better structure I have multiple .adoc files included in the index.adoc. At the end Jenkins should deploy the maven site to a web server when a change is detected.

But when I run "mvn clean site-deploy" I get an error in the modules when the site gets build:
`asciidoctor: WARNING: <stdin>: line 5: include file not found: /home/zao/devel/bug/doc/src/site/asciidoc/attributes.adoc`

The attributes.adoc is in the modules asciidoc folder. But asciidoctor searching it in the asciidoc folder from the parent project and not inside the module directory.

I figured out that when I do not use "clean" and I made changes at the index.adoc file it is working as expected. To reproduce this behavior I run:
`mvn clean site-deploy`
`touch module/src/site/asciidoc/index.adoc`
`mvn site-deploy`

Then everything is like expected. But this workaround is not nice to maintain with Jenkins and automated deploys.

I have a simple sample project where you can reproduce this behavior:
https://github.com/entfrickler/mvn-site-asciidoctor-test/tree/master

I have a workaround for those who have the same Problem in another branch:
https://github.com/entfrickler/mvn-site-asciidoctor-test/tree/workaround

In the workaround I use an absolute path which I get from the maven project.basedir and add the relative paths to the files to include. But this isn't a nice solution too, because I use the asciidoctor maven plugin without the site integration to write the documents and get a fast preview. The project.basedir, when I build it with "mvn site-deploy" is in every module the path to the parent. With "mvn compile" it points to the module path. So the workaround looks like:
`ifeval::["{build-site}" == "true"]`
`:asciidoc-dir: {project-basedir}/module/src/site/asciidoc`
`endif::[]`
`ifndef::asciidoc-dir[:asciidoc-dir: {project-basedir}/src/site/asciidoc]`
The {build-site} and {project-basedir} is set inside the pom.xml as an attribute for the asciidoctor maven plugin.

Is this a bug or have I something wrong in my project configuration?
Hopefully there is an easy way to fix this!

**Comment by abelsromero** [comment]
Hi,

I am going to need some time to delve into this one. One thing I can confirm that is obviously wrong is an issue with paths. When running with `clean` I get this:
```
asciidoctor: WARNING: <stdin>: line 5: include file not found: C:/home/personal/doc/src/site/asciidoc/global-attributes.adoc
asciidoctor: WARNING: <stdin>: line 5: include file not found: C:/home/personal/github/mvn-site-asciidoctor-test/src/site/asciidoc/attributes.adoc
asciidoctor: ERROR: <stdin>: line 7: only book doctypes can contain level 0 sections
asciidoctor: WARNING: <stdin>: line 28: include file not found: C:/home/personal/github/mvn-site-asciidoctor-test/src/site/asciidoc/debug.adoc
```
First message says `C:/home/personal/doc/`, which does not exist, and my project is cloned in `C:/home/personal/github/mvn-site-asciidoctor-test/`.

I'll try to have a look tomorrow.

> In the workaround I use an absolute path which I get from the maven project.basedir and add the relative paths to the files to include. But this isn't a nice solution too, because I use the asciidoctor maven plugin without the site integration to write the documents and get a fast preview.

Regarding the workaround is perfectly legal and normal, I use things like that also. Keep in mind that the process of editing and rendering are two separate phases, some similar to coding workflows than document writing with a WYSIWYG. I know your feeling and I too try to find common ground between those two phases.
I understand you render with the maven plugin to see the results. For a nicer experience I'd recommend using and editor with live preview. I use IntelliJ community and it works great (or you can use Atom.io or Brackets). Intellij it adds an attribute to the preview that allows to customize aspects only for preview purposes, for example changing icons in admonitions `ifndef::env-idea[:icons: font]`.

Another thing, I saw one thing in the example that I think it's not legal AsciiDoc, even if it 'works' and you get a result.
You are using an include to define attributes before the title, but the title should be the first element of the .adoc. Then you can define attributes there, but not with an include afaik. If you want to pass global attributes you should use the `<attributes>` section of the pom.

**Comment by entfrickler** [comment]
Thanks for the fast reply. I will try your hints according to the editor and will move the first include behind the title to have a stable build over version changes.

For me right now I can live with the workaround. But I think for new users it is not intuitive to do so. It took me some hours to understand what is going wrong. Because at the beginning I was confused about sometimes it was working and sometimes not... Than I figured out that when I change something in the main .adoc and build without a clean I get an result... For a while I didn't see that asciidoctor where looking in the wrong directory because the paths from the parent and the modules are quite similar...

**Comment by abelsromero** [comment]
I revisited this and the issue where submodules where not picking the correct base_dir is fixed (with latests v2.1.0). I cannot honestly say what fixed it though, there've been several refactors in that area other time.

Said that, I found the example had a couple of errors:
* Attributes sections had some linebreaks.
* `parentdir` definition in `module/.../attributes.adoc` has a invalid `doc` (`ifndef::parentdir[:parentdir: ../../../../../doc/src/site]`), correct path is `ifndef::parentdir[:parentdir: ../../../../src/site]`.

With those corrected everything works fine on first and clean executions.
